### PR TITLE
Restructure ODSP connection telemetry

### PR DESCRIPTION
**Pull request by ChumpChief**
I was finding the promise chaining and nesting a bit hard to follow, and the telemetry as well.  This change restructures the connection to unify connection logic and telemetry on AFD vs. non-AFD.  The distinction of cache vs. first-attempt vs. fallback should be derivable from telemetry without specially-named events.  This should also make it easier to observe connection timing (some of the flows don't log enough data to determine this currently).

This change does include one functional change beyond telemetry - currently we only write the AFD-valid cache if there wasn't a valid cache.  With this change, we'll write an updated timestamp to the cache on successful AFD connection regardless of the prior state of the cache.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspDocumentService.ts [hunk 1]
```diff
@@ -40,6 +40,59 @@ import { TokenFetchOptions } from "./tokenFetch";
...
+    return false;
+}
+
+/**
+ * Test if we deal with NetworkErrorBasic object and if it has enough information to make a call
+ * If in doubt, allow retries
+ *
+ * @param error - error object
+ */
+function canRetryOnError(error: any) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentService.ts:73]: similar API exists in delta manager. Can we move it somewhere such that all layers use same logic (and it is exported, such that consumers of our repo can use too?)

File: packages/drivers/odsp-driver/src/odspDocumentService.ts [hunk 2]
```diff
@@ -40,6 +40,59 @@ import { TokenFetchOptions } from "./tokenFetch";
...
+ *
+ * @param key - localStorage key
+ * @returns whether or not the write succeeded
+ */
+function writeLocalStorage(key: string, value: string) {
+    try {
+        localStorage.setItem(key, value);
+        return true;
+    } catch (e) {
+        debug(`Could not write to localStorage due to ${e}`);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentService.ts:91]: I'd remove debug(). These statements are great for local debugging, but should be removed or converted to telemetry in shipping product. The ideal long term direction - everything is in telemetry, but verboseness is controlled (locally & remotely depending on triggers).

File: packages/drivers/odsp-driver/src/odspDocumentService.ts [hunk 3]
```diff
@@ -288,157 +308,126 @@ export class OdspDocumentService implements IDocumentService {
...
-                    logger.sendPerformanceEvent({
-                        eventName: "FailedAfdUrl-NoNonAfdFallback",
-                    }, connectionError);
-                }
+            } catch (connectionError) {
+                // Log before throwing
+                const canRetry = canRetryOnError(connectionError);
+                this.logger.sendTelemetryEvent(
+                    {
+                        eventName: "NonAfdConnectionFail",
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentService.ts:348]: Since you are touching this code :)
I'd prefer different approach, The one where connectToDeltaStreamWithRetry() collects write data and stucks it on error object and throws it, without logging anything. And it's up to DeltaManager to reason about log or not to log it.
The reason I like it better is that we have central place to optimize amount of telemetry. I.e. today, for example, it will log error only of first try in a loop of retries, and this is good for offline scenario, where we keep retrying every 8 seconds, accumulating ton of telemetry that Aria flushes when client gets connected.

This does not solve telemetry needs for successful connection - I think it's fine for this code to issue an event in such cases with data on what was tried etc - number of these events will be much smaller than number of failures in disconnected state.

File: packages/drivers/odsp-driver/src/odspDocumentService.ts [hunk 4]
```diff
@@ -288,157 +308,126 @@ export class OdspDocumentService implements IDocumentService {
...
+                throw connectionError;
+            }
+        };
+
+        const afdCacheValid = isAfdCacheValid();
+
+        // First use the AFD URL if we've logged a successful AFD connection in the cache - its presence in the cache
+        // means the non-AFD url has failed in the past, in which case we would prefer to skip doing another
+        // attempt->fail on the non-AFD.
+        if (afdCacheValid && afdUrl !== undefined) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentService.ts:391]: I'd personally change it. I'd rather create a "plan" of connections in the form of uris: [string], and then just execute it.
I'm not a fan of so many telemetry events & verboseness - I'd rather see error object (or error properties object) being populated as we populate the plan, execute the plan, and then being thrown at the end.

**Review by vladsud — APPROVED** [review]
Like the directions, and it's definitely and improvement, but left some comments on telemetry and impact on size of it.

File: packages/drivers/odsp-driver/src/odspDocumentService.ts (hunk 1 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentService.ts]: Moved to driver-utils in latest

File: packages/drivers/odsp-driver/src/odspDocumentService.ts (hunk 2 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentService.ts]: Done in latest.

File: packages/drivers/odsp-driver/src/odspDocumentService.ts (hunk 4 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentService.ts:393]: Latest change reduces emissions to one per connection attempt, emitted when the attempt succeeds or fails.  Since we don't have a way to emit multiple events with a single telemetry submission, I think this is minimal verbosity while still achieving the goal of this PR of rounding out the missing telemetry.

File: packages/drivers/odsp-driver/src/odspDocumentService.ts (hunk 3 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentService.ts:336]: I'd prefer not to attempt this in this PR.  Doing this in a manner that leaves DeltaManager agnostic to the particular driver likely requires establishing some new interfaces which is beyond the scope of this PR's goal, which is to get improved telemetry up and running to help with investigating OCE issue.  I'm also not entirely sure I like the idea of splitting the telemetry emission logic between layers of the stack like this.

File: packages/drivers/odsp-driver/src/odspDocumentService.ts [hunk 5]
```diff
@@ -288,157 +296,124 @@ export class OdspDocumentService implements IDocumentService {
...
+                    documentId,
+                    token,
+                    io,
+                    client,
+                    nonAfdUrl,
+                    20000,
+                    this.logger,
+                );
+                const endTime = performance.now();
+                this.logger.sendPerformanceEvent({
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentService.ts:325]: Nit: should we think about some wrapper like PerformanceEvent.timedExecAsync() for simple cases like this? I.e. same as timedExecAsync but without start event? Maybe an optional argument for timedExecAsync.
catch() blocks makes it a bit tricky and code would be rather similar to what it is today, but it feels like it decreases chances of developers to screwed up telemetry.

File: packages/drivers/odsp-driver/src/odspDocumentService.ts [hunk 6]
```diff
@@ -288,157 +296,124 @@ export class OdspDocumentService implements IDocumentService {
...
-                    logger.sendPerformanceEvent({
-                        eventName: "FailedAfdUrlFallback",
-                        duration: endNonAfd - startNonAfd,
-                    }, retryError);
-                    throw retryError;
+                );
+                // Set the successful connection attempt in the cache so we can skip the non-AFD failure the next time
+                // we try to connect and immediately try AFD instead.
+                writeLocalStorage(lastAfdConnectionTimeMsKey, Date.now().toString());
+                const endTime = performance.now();
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentService.ts:364]: I'd move it up one line - measure only network (as much as it's possible) - who knows, maybe writing to localStorage may (?) result in some user prompt

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentService.ts (hunk 5 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentService.ts:325]: I'm not really a fan of `timedExecAsync` since it shows up on the stack.  One thing I did experiment with was using `performance.mark()`, `performance.measure()`, and `PerformanceObserver` to standardize telemetry measuring plus also add performance trace data in dev tools.  Maybe something like adding `startMeasure()` + `endMeasure()` helpers.  I decided that probably deserves more in-depth thought than I want to put into this PR though.

File: packages/drivers/odsp-driver/src/odspDocumentService.ts (hunk 6 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentService.ts]: Good catch, I missed this one when I was moving the others up.  Fixed in latest

**Review by vladsud — APPROVED** [review]
